workspaces/cicd-statistics/plugins/cicd-statistics-module-buildkite/src/api/buildkite.ts

```diff
@@ -0,0 +1,145 @@
+  CicdState,
+  CicdConfiguration,
+  CicdDefaults,
+  FetchBuildsOptions,
+  FilterStatusType,
+} from '@backstage-community/plugin-cicd-statistics';
+import { Entity } from '@backstage/catalog-model';
+import { toBuilds, buildStatusMap } from './utils';
+import { BuildkiteApi } from '@roadiehq/backstage-plugin-buildkite';
+
+/**
+ * BuildkiteClient represents an initialized Buildkite client.
+ *
+ * @public
+ */
+export type BuildkiteClient = {
+  /* the actual API */
+  api: InstanceType<typeof BuildkiteApi>;
+
+  /* the Buildkite organization; retrieved from the entity */
+  org: string;
+
+  /* the Buildkite pipeline name; retrieved from the entity */
+  pipeline: string;
+};
+
+/**
+ * CicdStatisticsApiBuildkiteOpts configures a new CicdStatisticsApiBuildkite.
+ *
+ * @public
+ */
+export type CicdStatisticsApiBuildkiteOpts = {
+  /* the DiscoveryApi to use */
+  discoveryApi: DiscoveryApi;
+
+  /* the FetchApi to use */
+  fetchApi: FetchApi;
+
+  /* the CicdDefaults to use */
+  cicdDefaults?: Partial<CicdDefaults>;
+
+  /* the path to the Buildkite API Backstage backend proxy endpoint */
+  proxyPath?: string;
+};
+
+/**
+ * Extracts the CI/CD statistics from a Buildkite pipeline.
+ *
+ * @public
+ */
+export class CicdStatisticsApiBuildkite implements CicdStatisticsApi {
+  readonly #cicdDefaults: Partial<CicdDefaults>;
+  readonly #discoveryApi: DiscoveryApi;
+  readonly #fetchApi: FetchApi;
+  readonly #proxyPath?: string;
+
+  constructor(opts: CicdStatisticsApiBuildkiteOpts) {
+    this.#cicdDefaults = opts.cicdDefaults ? opts.cicdDefaults : {};
+    this.#discoveryApi = opts.discoveryApi;
+    this.#fetchApi = opts.fetchApi;
+    this.#proxyPath = opts.proxyPath;
+  }
+
+  public async createBuildkiteApi(entity: Entity): Promise<BuildkiteClient> {
+    const pipelineAnnotation =
+      entity.metadata.annotations?.['buildkite.com/pipeline'] || '';
+    const pipelineParts = pipelineAnnotation.split('/');
+    const org = pipelineParts[0];
+    const pipeline = pipelineParts[1] ? pipelineParts[1] : '';
+
+    return {
+      api: new BuildkiteApi({
+        discoveryApi: this.#discoveryApi,
+        fetchApi: this.#fetchApi,
+        proxyPath: this.#proxyPath,
+      }),
+      org,
+      pipeline,
+    };
+  }
+
+  public async fetchBuilds(options: FetchBuildsOptions): Promise<CicdState> {
+    const {
+      entity,
+      updateProgress,
+      filterStatus = ['all'],
+      filterType = 'all',
+    } = options;
+    const { api, org, pipeline } = await this.createBuildkiteApi(entity);
+    updateProgress(0, 0, 0);
+
+    // Note the cicd-statistics plugin regards 'master' as a generic term
+    // referring to a particular default 'trunk' branch, and not necessarily
+    // the literal 'master' branch name.
+    const branch =
+      filterType === 'master'
+        ? await this.getDefaultBranch(api, org, pipeline)
+        : undefined;
+    const bkBuilds = await api.getBuilds(org, pipeline, 0, 25, branch);
```

Comment:
Also consider wrapping API calls in try-catch blocks to handle API failures more gracefully.

Reply:
@04kash I believe that error handling already exists at a more appropriate higher level, [within the frontend](https://github.com/backstage/community-plugins/blob/main/workspaces/cicd-statistics/plugins/cicd-statistics/src/hooks/use-cicd-statistics.ts#L123), no?

Plus, even if there's additional `try`/`catch` handling here within `fetchBuilds`, [the caller of `fetchBuilds`](https://github.com/backstage/community-plugins/blob/main/workspaces/cicd-statistics/plugins/cicd-statistics/src/hooks/use-cicd-statistics.ts#L123) still needs its own `try`/`catch` handling, and so I'd argue adding additional `try`/`catch` logic here only serves to add complexity and risk obfuscating underlying errors. Do you disagree?

(And, also note that the underlying Buildkite API client [already has its own handling of not-OK HTTP responses from the Buildkite API](https://github.com/RoadieHQ/roadie-backstage-plugins/blob/main/plugins/frontend/backstage-plugin-buildkite/src/api/index.ts#L86-L90), which ensures such errors are adequately surfaced & handled by [the frontend](https://github.com/backstage/community-plugins/blob/main/workspaces/cicd-statistics/plugins/cicd-statistics/src/hooks/use-cicd-statistics.ts#L123)).

Reply:
Additionally, for whatever it's worth, the Buildkite `fetchBuilds` implementation pattern seen here emulates & maintains consistency with [the existing pattern implemented by the cicd-statistics-module-gitlab](https://github.com/mdb/community-plugins/blob/mdb%2Fcicd-stats-buildkite/workspaces/cicd-statistics/plugins/cicd-statistics-module-gitlab/src/api/gitlab.ts#L111-L149).

Reply:
Alright, thanks for clarifying.